lib/src/bitarray.dart

```diff
@@ -0,0 +1,317 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection' show ListMixin;
+import 'dart:typed_data' show Uint32List;
+import 'package:collection/collection.dart' show NonGrowableListMixin;
+
+/// A BitArray implementation for Dart to store boolean values.
+///
+/// Uses list of integers as internal storage to reduce memory usage.
+/// Implemented as fixed-length [List] of [bool].
+/// Throws error when calling any operation with length modifying.
+///
+/// Support several ways to perform logical operations (NOT, OR, AND, XOR):
+/// * instance methods (mutate [this]): [BitArray.not], [BitArray.or], [BitArray.and], [BitArray.xor];
+/// * producing new instance of [BitArray]:
+///     * logical operators: [BitArray.~], [BitArray.|], [BitArray.&], [BitArray.^].
+///     * logical gates (as static methods): [BitArray.notGate], [BitArray.orGate], [BitArray.andGate], [BitArray.xorGate].
+/// All arguments for not-unary logical operations must have the same [length] or [ArgumentError] will be thrown.
+class BitArray with ListMixin<bool>, NonGrowableListMixin<bool> {
+  int _length;
+
+  Uint32List _data;
+
+  /// Mask to set tail bits to zero.
+  int _tailMask;
+
+  static const List<int> _numOfBitsLookup = [
+    0, 1, 1, 2, 1, 2, 2, 3, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    5, 6, 6, 7, 6, 7, 7, 8
+  ];
+
+  /// Creates a [BitArray] with [length].
+  ///
+  /// All initial values are [false], set [filled] to [true] to make them [true].
+  BitArray(int length, [bool filled = false]) {
+    ArgumentError.checkNotNull(length);
+    RangeError.checkNotNegative(length, 'length');
+
+    _length = length;
```

Comment:
Use initializers rather than assignment.
If this is difficult, have a private constructor `BitList._(this._length, this._data, this._tailMask);` and make this constructor be a factory constructor.